source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,992 @@
+    AzureIoTProvisioningClientResult_t xReturn;
+    uint32_t ulWaitTime;
+
+    do
+    {
+        if( xProvClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_COMPLETE )
+        {
+            AZLogDebug( ( "AzureIoTProvisioning state is already in complete state" ) );
+            break;
+        }
+
+        if( ulTimeoutMilliseconds > azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS )
+        {
+            ulTimeoutMilliseconds -= azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS;
+            ulWaitTime = azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS;
+        }
+        else
+        {
+            ulWaitTime = ulTimeoutMilliseconds;
+            ulTimeoutMilliseconds = 0;
+        }
+
+        prvProvClientTriggerAction( xProvClientHandle );
+
+        if( ( xResult =
+                  AzureIoTMQTT_ProcessLoop( &( xProvClientHandle->_internal.xMQTTContext ),
+                                            ulWaitTime ) ) != eAzureIoTMQTTSuccess )
+        {
+            AZLogError( ( "AzureIoTProvisioning  failed to process loop: ProcessLoopDuration=%u, Error=%d",
+                          ulTimeoutMilliseconds, xResult ) );
+            prvProvClientUpdateState( xProvClientHandle, eAzureIoTProvisioningFailed );
+            break;
+        }
+    } while( ulTimeoutMilliseconds );
+
+    if( ( xProvClientHandle->_internal.ulWorkflowState != azureiotprovisioningWF_STATE_COMPLETE ) )
+    {
+        xReturn = eAzureIoTProvisioningPending;
+    }
+    else
+    {
+        xReturn = xProvClientHandle->_internal.ulLastOperationResult;
+    }
+
+    return xReturn;
+}
+/*-----------------------------------------------------------*/
+
+
+/**
+ *  Process MQTT Subscribe Ack from Provisioning Service
+ *
+ */
+static void prvProvClientMQTTProcessSubAck( AzureIoTProvisioningClientHandle_t xProvClientHandle,
+                                            AzureIoTMQTTPublishInfo_t * pxPacketInfo )
+{
+    ( void ) pxPacketInfo;
+
+    prvProvClientUpdateState( xProvClientHandle, eAzureIoTProvisioningSuccess );
+}
+/*-----------------------------------------------------------*/
+
+/**
+ * Process MQTT Response from Provisioning Service
+ *
+ */
+static void prvProvClientMQTTProcessResponse( AzureIoTProvisioningClientHandle_t xProvClientHandle,
+                                              AzureIoTMQTTPublishInfo_t * pxPublishInfo )
+{
+    if( xProvClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_REQUESTING )
+    {
+        if( ( pxPublishInfo->xPayloadLength + pxPublishInfo->usTopicNameLength ) <=
+            sizeof( xProvClientHandle->_internal.ucProvisioningLastResponse ) )
+        {
+            /* Copy topic + payload into the response buffer.
+             * ucProvisioningLastResponse = [ topic payload ] */
+            xProvClientHandle->_internal.usLastResponseTopicLength =
+                pxPublishInfo->usTopicNameLength;
+            memcpy( xProvClientHandle->_internal.ucProvisioningLastResponse,
```

Comment:
MQTT stack owns the memory and span depends does not take the ownership. So that is the reason middleware copy to memory that it owns and access it outside mqtt callback.